kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/InputStreamPumper.java

```diff
@@ -15,76 +15,84 @@
  */
 package io.fabric8.kubernetes.client.utils;
 
-import io.fabric8.kubernetes.client.Callback;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.io.Closeable;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.InputStreamReader;
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
 
-@Deprecated
-// This is going to be an abstract class in the future.
-// Use NonBlockingInputStreamPumper or BlockingInputStreamPumper instead.
-public class InputStreamPumper implements Runnable, Closeable {
+public class InputStreamPumper {
+  
+  private InputStreamPumper() {
+  }
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(InputStreamReader.class);
+  private static final Logger LOGGER = LoggerFactory.getLogger(InputStreamPumper.class);
 
-    final InputStream in;
-    final Callback<byte[]> callback;
-    final Runnable onClose;
-    final AtomicBoolean closed = new AtomicBoolean(false);
+  public interface Writable {
 
-    volatile boolean keepReading = true;
-    Thread thread;
+    void write(byte[] b, int off, int len) throws IOException;
 
-    public InputStreamPumper(InputStream in, Callback<byte[]> callback) {
-      this(in, callback, null);
-    }
+  }
 
-    public InputStreamPumper(InputStream in, Callback<byte[]> callback, Runnable onClose) {
-        this.in = in;
-        this.callback = callback;
-        this.onClose = onClose;
-    }
+  /**
+   * Relies on {@link InputStream#available()} and a Thread sleep to ensure that the reads are interruptible.
+   */
+  public static InputStream asInterruptible(InputStream is) {
+    return new InputStream() {
 
-    @Override
-    public void run() {
-        synchronized (this) {
-          thread = Thread.currentThread();
-        }
-        byte[] buffer = new byte[1024];
-        try {
-          int length;
-          while (keepReading && !Thread.currentThread().isInterrupted() && (length = in.read(buffer)) != -1) {
-                byte[] actual = new byte[length];
-                System.arraycopy(buffer, 0, actual, 0, length);
-                callback.call(actual);
-              }
-        } catch (IOException e) {
-          if (!keepReading) {
-            return;
+      @Override
+      public int read() throws IOException {
+        throw new UnsupportedOperationException();
+      }
+      
+      @Override
+      public int read(byte[] b, int off, int len) throws IOException {
+        while (!Thread.currentThread().isInterrupted()) {
+          if (is.available() > 0) {
+            return is.read(b, off, len);
           }
-          if (!thread.isInterrupted()) {
-            LOGGER.error("Error while pumping stream.", e);
-          } else {
-            LOGGER.debug("Interrupted while pumping stream.");
+          try {
+            Thread.sleep(50);
```

Comment:
I like the idea of using `available()` which is non-blocking.
I have mixed feelings though about the `Thread.sleep(50)` and more specifically about the `50`.
How did you came up with it? Is it something generally recommended? If so, I'd like a pointer (and maybe as a comment too). If not, do you think it would make sense to try and have something configurable or smarter?

Reply:
It's just what was in the existing NonBlockingInputStreamPumper.   It would be fine to add a parameter to the method asInterruptable that was the wait time.

Reply:
just added a comment that the 50 was from the previous code.

Reply:
Ooops! I didn't realize that. Sorry!